File: sdks/dotnet/src/ModelDriveProtocol.Client/PathUtils.cs
using System.Text.RegularExpressions;

namespace ModelDriveProtocol.Client;

public sealed record PathPatternMatch(
    IReadOnlyDictionary<string, object?> Params,
    IReadOnlyList<int> Specificity);

public static partial class PathUtils
{
    private static readonly HashSet<string> ReservedLeafNames = new(StringComparer.Ordinal)
    {
        "skill.md",
        "prompt.md",
        "SKILL.md",
        "PROMPT.md"
    };

    public static bool IsPathPattern(string value) => ValidatePath(value, allowParams: true);

    public static bool IsConcretePath(string value) => ValidatePath(value, allowParams: false);

    public static bool IsSkillPath(string value)
    {
        if (!IsPathPattern(value))
        {
            return false;
        }

        string[] segments = SplitPath(value);
        return segments.Length > 0 && (segments[^1] == "skill.md" || segments[^1] == "SKILL.md");
    }

    public static bool IsPromptPath(string value)
    {
        if (!IsPathPattern(value))
        {
            return false;
        }

        string[] segments = SplitPath(value);
        return segments.Length > 0 && (segments[^1] == "prompt.md" || segments[^1] == "PROMPT.md");
    }

    public static PathPatternMatch? MatchPathPattern(string pattern, string path)
    {
        if (!IsPathPattern(pattern) || !IsConcretePath(path))
        {
            return null;
        }

        string[] patternSegments = SplitPath(pattern);
        string[] pathSegments = SplitPath(path);
        if (patternSegments.Length != pathSegments.Length)
        {
            return null;
        }

        Dictionary<string, object?> parameters = new(StringComparer.Ordinal);
        List<int> specificity = new(patternSegments.Length);

        for (int index = 0; index < patternSegments.Length; index += 1)
        {
            string patternSegment = patternSegments[index];
            string pathSegment = pathSegments[index];

            if (patternSegment.StartsWith(':'))
            {
                parameters[patternSegment[1..]] = pathSegment;
                specificity.Add(0);
                continue;
            }

            if (!string.Equals(patternSegment, pathSegment, StringComparison.Ordinal))
            {
                return null;
            }

            specificity.Add(IsReservedLeafName(patternSegment) ? 1 : 2);
        }

        return new PathPatternMatch(parameters, specificity);
    }

    public static int ComparePathSpecificity(IReadOnlyList<int> left, IReadOnlyList<int> right)
    {
        int maxLength = Math.Max(left.Count, right.Count);
        for (int index = 0; index < maxLength; index += 1)
        {
            int leftValue = index < left.Count ? left[index] : -1;
            int rightValue = index < right.Count ? right[index] : -1;
            if (leftValue != rightValue)
            {
                return leftValue.CompareTo(rightValue);
            }
        }
        return 0;
    }

    private static bool ValidatePath(string value, bool allowParams)
    {
        if (string.IsNullOrWhiteSpace(value) || !value.StartsWith('/') || value.Contains('?') || value.Contains('#'))
        {
            return false;
        }

        string[] segments = SplitPath(value);
        if (segments.Length == 0)
        {
            return false;
        }

        for (int index = 0; index < segments.Length; index += 1)
        {
            string segment = segments[index];
            bool isLast = index == segments.Length - 1;
            if (string.IsNullOrEmpty(segment))
            {
                return false;
            }
            if (IsReservedLeafName(segment))
            {
                if (!isLast)
                {
                    return false;
                }
                continue;
            }
            if (allowParams && ParameterSegment().IsMatch(segment))
            {
                continue;
            }
            if (!StaticSegment().IsMatch(segment))
            {
                return false;
            }
        }

        return true;
    }

    private static string[] SplitPath(string value) =>
        value.TrimStart('/').Split('/');

    private static bool IsReservedLeafName(string value) => ReservedLeafNames.Contains(value);

    [GeneratedRegex("^(?:[a-z0-9](?:[a-z0-9_-]*)?|\\.[a-z0-9](?:[a-z0-9_-]*)?)$")]
    private static partial Regex StaticSegment();

    [GeneratedRegex("^:[a-z0-9](?:[a-z0-9_-]*)$")]
    private static partial Regex ParameterSegment();
}

File: sdks/jvm/java-client/src/main/java/io/modeldriveprotocol/client/PathUtils.java
package io.modeldriveprotocol.client;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

final class PathUtils {
  private static final Pattern STATIC_SEGMENT_PATTERN =
      Pattern.compile("^(?:[a-z0-9](?:[a-z0-9_-]*)?|\\.[a-z0-9](?:[a-z0-9_-]*)?)$");
  private static final Pattern PARAM_SEGMENT_PATTERN =
      Pattern.compile("^:[a-z0-9](?:[a-z0-9_-]*)$");
  private static final Set<String> RESERVED_LEAF_NAMES =
      Set.of("skill.md", "prompt.md", "SKILL.md", "PROMPT.md");

  private PathUtils() {}

  static boolean isPathPattern(String value) {
    return validatePath(value, true);
  }

  static boolean isConcretePath(String value) {
    return validatePath(value, false);
  }

  static boolean isSkillPath(String value) {
    List<String> segments = splitPath(value);
    if (!isPathPattern(value) || segments.isEmpty()) {
      return false;
    }
    String leaf = segments.get(segments.size() - 1);
    return "skill.md".equals(leaf) || "SKILL.md".equals(leaf);
  }

  static boolean isPromptPath(String value) {
    List<String> segments = splitPath(value);
    if (!isPathPattern(value) || segments.isEmpty()) {
      return false;
    }
    String leaf = segments.get(segments.size() - 1);
    return "prompt.md".equals(leaf) || "PROMPT.md".equals(leaf);
  }

  static PathPatternMatch matchPathPattern(String pattern, String path) {
    if (!isPathPattern(pattern) || !isConcretePath(path)) {
      return null;
    }

    List<String> patternSegments = splitPath(pattern);
    List<String> pathSegments = splitPath(path);
    if (patternSegments.size() != pathSegments.size()) {
      return null;
    }

    Map<String, Object> params = new LinkedHashMap<>();
    List<Integer> specificity = new ArrayList<>();
    for (int index = 0; index < patternSegments.size(); index += 1) {
      String patternSegment = patternSegments.get(index);
      String pathSegment = pathSegments.get(index);
      if (patternSegment.startsWith(":")) {
        params.put(patternSegment.substring(1), pathSegment);
        specificity.add(0);
        continue;
      }

      if (!patternSegment.equals(pathSegment)) {
        return null;
      }

      specificity.add(RESERVED_LEAF_NAMES.contains(patternSegment) ? 1 : 2);
    }

    return new PathPatternMatch(params, specificity);
  }

  static int comparePathSpecificity(List<Integer> left, List<Integer> right) {
    int maxLength = Math.max(left.size(), right.size());
    for (int index = 0; index < maxLength; index += 1) {
      int leftValue = index < left.size() ? left.get(index) : -1;
      int rightValue = index < right.size() ? right.get(index) : -1;
      if (leftValue != rightValue) {
        return leftValue - rightValue;
      }
    }
    return 0;
  }

  private static boolean validatePath(String value, boolean allowParams) {
    if (value == null || !value.startsWith("/") || value.contains("?") || value.contains("#")) {
      return false;
    }

    List<String> segments = splitPath(value);
    if (segments.isEmpty()) {
      return false;
    }

    for (int index = 0; index < segments.size(); index += 1) {
      String segment = segments.get(index);
      if (segment.isEmpty()) {
        return false;
      }

      boolean isLast = index == segments.size() - 1;
      if (RESERVED_LEAF_NAMES.contains(segment)) {
        if (!isLast) {
          return false;
        }
        continue;
      }

      if (allowParams && PARAM_SEGMENT_PATTERN.matcher(segment).matches()) {
        continue;
      }

      if (!STATIC_SEGMENT_PATTERN.matcher(segment).matches()) {
        return false;
      }
    }

    return true;
  }

  private static List<String> splitPath(String value) {
    String[] segments = value.split("/");
    List<String> result = new ArrayList<>();
    for (int index = 1; index < segments.length; index += 1) {
      result.add(segments[index]);
    }
    return result;
  }

  record PathPatternMatch(Map<String, Object> params, List<Integer> specificity) {}
}
